File: internal/web/records.go
package web

import (
	"encoding/json"
	"net/http"
	"path/filepath"
	"strings"
	"time"

	"github.com/ZachCurry13/isoshelf/internal/settings"
	"github.com/ZachCurry13/isoshelf/internal/state"
)

// Where a folder's records are kept is the one setting that owns files. The
// rest of Settings is switches; this one has to pick up what isoshelf has
// learned about the folder and put it down somewhere else, with nothing lost
// on the way and nothing deleted that the user didn't choose to lose. So it
// asks first (planRecords), and says afterwards what it did (setRecords).

// recordsJSON is what the page shows and sends back.
type recordsJSON struct {
	// Location is settings.InFolder, WithApp or Elsewhere.
	Location string `json:"location"`
	// Dir is the folder for Elsewhere.
	Dir string `json:"dir,omitempty"`
	// File is where this folder's records are right now, so Settings can
	// point at the actual file rather than describing it.
	File string `json:"file,omitempty"`
	// Moved is what a move just did. Only the answer to that move carries
	// it; the page keeps it for as long as it goes on saying so.
	Moved *movedJSON `json:"moved,omitempty"`
}

// movedJSON is what moving the records did, or would do.
type movedJSON struct {
	// What is "move"; "keep", when both places had records and the ones
	// already where they were going win; "use", when only that place had
	// any; "none", when nothing has been saved about the folder yet; or
	// "same", when nothing changes.
	What      string    `json:"what"`
	From      string    `json:"from"`
	To        string    `json:"to"`
	FromDir   string    `json:"from_dir"`
	ToDir     string    `json:"to_dir"`
	FromSaved time.Time `json:"from_saved,omitzero"`
	ToSaved   time.Time `json:"to_saved,omitzero"`
}

func movedFor(m state.Moved) *movedJSON {
	out := &movedJSON{
		From: m.From, To: m.To, FromDir: filepath.Dir(m.From), ToDir: filepath.Dir(m.To),
		FromSaved: m.FromSaved, ToSaved: m.ToSaved,
	}
	switch {
	case m.From == m.To:
		out.What = "same"
	case m.Moved:
		out.What = "move"
	case m.Kept:
		out.What = "keep"
	case !m.ToSaved.IsZero():
		out.What = "use"
	default:
		out.What = "none"
	}
	return out
}

// recordsLocked describes where the open folder's records are. s.mu must be
// held.
func (s *Server) recordsLocked(saved settings.Settings) recordsJSON {
	if s.target == "" {
		return recordsJSON{Location: settings.InFolder}
	}
	r := saved.RecordsFor(s.target)
	out := recordsJSON{Location: r.Location, Dir: r.Dir}
	home := state.Home(saved.RecordsHome(s.target, s.cfg.Dirs.Config))
	if file, err := home.File(s.target); err == nil {
		out.File = file
	}
	return out
}

// recordsChange is a request to keep the open folder's records somewhere
// else, worked out: the folder, where its records are, where they would go,
// and the settings that would say so.
type recordsChange struct {
	target   string
	from, to state.Home
	after    settings.Settings
}

// readRecordsChange reads what the page asked for and checks that it can be
// done. Asking and doing both come through here, so nobody is asked a
// question whose answer would then be refused. It writes any refusal itself.
func (s *Server) readRecordsChange(w http.ResponseWriter, r *http.Request) (recordsChange, bool) {
	var req struct {
		Location string `json:"location"`
		Dir      string `json:"dir"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "bad request")
		return recordsChange{}, false
	}
	choice := settings.CleanRecordsLocation(req.Location)
	if choice == "" {
		writeError(w, http.StatusBadRequest, "Choose where to keep this folder's records.")
		return recordsChange{}, false
	}

	// Nothing may be reading or writing the records while they move: a scan
	// or a download that saved afterwards would save to the old place, and
	// its work would be the part that went missing.
	s.mu.Lock()
	target, busy := s.target, s.busyLocked()
	s.mu.Unlock()
	if target == "" {
		writeError(w, http.StatusBadRequest, "Open a folder first.")
		return recordsChange{}, false
	}
	if busy != "" {
		writeError(w, http.StatusConflict, busy)
		return recordsChange{}, false
	}

	want := settings.Records{Location: choice, Dir: req.Dir}
	if choice == settings.Elsewhere {
		home, err := state.CleanHome(req.Dir)
		if err != nil {
			writeError(w, http.StatusBadRequest, err.Error())
			return recordsChange{}, false
		}
		want.Dir = string(home)
		if under, err := isUnder(target, want.Dir); err == nil && under {
			writeError(w, http.StatusBadRequest,
				"That folder is inside the one isoshelf is looking after, so its records would be on the same drive as the images. Keep them in the folder instead.")
			return recordsChange{}, false
		}
	}

	// Where they are now has to be worked out before the new answer is set:
	// after is a copy of saved, but the answers are a map, and a copied map
	// is the same map. Asked afterwards, saved would already say the new place.
	saved := s.loadSettings()
	from := state.Home(saved.RecordsHome(target, s.cfg.Dirs.Config))
	after := saved
	after.SetRecordsFor(target, want)
	return recordsChange{
		target: target,
		from:   from,
		to:     state.Home(after.RecordsHome(target, s.cfg.Dirs.Config)),
		after:  after,
	}, true
}

// planRecords says what moving the open folder's records would do, and does
// none of it, so the page can ask before anything moves.
func (s *Server) planRecords(w http.ResponseWriter, r *http.Request) {
	c, ok := s.readRecordsChange(w, r)
	if !ok {
		return
	}
	plan, err := state.Plan(c.from, c.to, c.target)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Couldn't look at this folder's records: "+err.Error())
		return
	}
	writeJSON(w, http.StatusOK, movedFor(plan))
}

// setRecords moves the open folder's records somewhere else, remembers that
// that is where they now are, and says what it did - including when it moved
// nothing, because records were already waiting where they were going.
func (s *Server) setRecords(w http.ResponseWriter, r *http.Request) {
	c, ok := s.readRecordsChange(w, r)
	if !ok {
		return
	}
	moved, err := state.Move(c.from, c.to, c.target)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Couldn't move this folder's records: "+err.Error())
		return
	}
	if err := settings.Save(s.cfg.Dirs.Config, c.after); err != nil {
		// The records are at the new place but nothing would look for them
		// there. Put them back rather than leave them stranded.
		state.Move(c.to, c.from, c.target)
		writeError(w, http.StatusInternalServerError, "Couldn't remember where the records went, so they were put back: "+err.Error())
		return
	}

	// Read them from where they now are, so the page shows what isoshelf will
	// actually use from here on.
	st, err := c.to.Load(c.target)
	s.mu.Lock()
	if err == nil && s.target == c.target {
		s.st = st
	}
	s.mu.Unlock()
	out := s.pageState()
	out.Records.Moved = movedFor(moved)
	writeJSON(w, http.StatusOK, out)
}

// isUnder reports whether dir is inside parent, or is parent itself. Rel
// gives an error for two different drives, which is a plain no.
func isUnder(parent, dir string) (bool, error) {
	rel, err := filepath.Rel(parent, dir)
	if err != nil {
		return false, err
	}
	return rel == "." || (rel != ".." && !strings.HasPrefix(rel, ".."+string(filepath.Separator))), nil
}

File: internal/web/static/records.js
"use strict";

// Where the open folder's records are kept. The one setting that owns files:
// changing it picks them up and puts them down somewhere else. So it says
// what would move, from where to where, before anything does - and afterwards
// what happened, including when records were already waiting and nothing
// moved at all. The maintainer had to ask what the dropdown did, after it had
// already done it; that was the bug.

// recordsChoice is the answer being chosen right now, while "a folder you
// pick" is showing its box and nothing has been saved yet. Null means the
// panel shows what isoshelf actually does.
let recordsChoice = null;

// recordsResult is what the last move did, and for which folder. It is said
// under the setting for as long as that folder stays open.
let recordsResult = null;

// recordsBusy is set while a change is being asked about or made. A second
// one meanwhile - Enter and the button together, say - is ignored: two
// questions share the one dialog, and the answer to the second would be lost.
let recordsBusy = false;

// chooseRecords is what the list and the folder box call. It asks isoshelf
// what the change would do, asks the user whenever records would move or a
// different set would be used, and only then makes the change.
async function chooseRecords(change) {
  if (recordsBusy) return;
  recordsBusy = true;
  try {
    let plan;
    try {
      plan = await api("POST", "/api/records/plan", change);
    } catch (err) {
      showNotice(err.message, true);
      recordsBack(change);
      return;
    }
    if (plan.what === "same") {
      recordsBack(change);
      return;
    }
    if (plan.what !== "none" && !(await askRecords(plan, change))) {
      recordsBack(change);
      return;
    }
    await saveRecords(change);
  } finally {
    recordsBusy = false;
  }
}

// saveRecords makes the change. It is not /api/settings: this one picks up
// files and puts them down somewhere else, so it has its own request, and it
// can say no - while a scan runs, or to a folder that would put the records
// on the same drive as the images.
async function saveRecords(change) {
  try {
    state = await api("POST", "/api/records", change);
    recordsResult = { ...state.records.moved, target: state.target };
    recordsChoice = null;
    render();
  } catch (err) {
    showNotice(err.message, true);
    recordsBack(change);
  }
}

// recordsBack puts the list back to what isoshelf actually does, after a no
// or a refusal. A typed folder stays in its box, one edit from being right.
function recordsBack(change) {
  if (change.location === "custom") return;
  recordsChoice = null;
  renderSettings(true);
}

// askRecords is the question before a change. It names both files and what
// stays behind - or, when records are already waiting where these would go,
// which set wins and how old each one is.
function askRecords(plan, change) {
  const cancel = { label: "Cancel", value: false };
  if (plan.what === "move") {
    // Kept away from the folder, they are found by its path - the one thing
    // about this choice that can surprise somebody later, so it is said now.
    const away = change.location === "folder" ? "" :
      " Kept away from the folder, they're found by its path, so the same drive at a " +
      "different letter or mount point starts with nothing.";
    return ask("Move this folder's records?",
      "Its history, the images you starred and what each file turned out to be move " +
      `from one file to the other. Your images, the archive and part-finished downloads ` +
      `stay in ${state.target}.` + away,
      [{ label: "Move them", value: true, primary: true }, cancel],
      el("dl", { class: "records-move" },
        el("dt", {}, "From"), el("dd", {}, plan.from),
        el("dt", {}, "To"), el("dd", {}, plan.to)));
  }
  const left = plan.what === "keep"
    ? ` The ones in ${plan.from_dir}, saved ${timeAgo(plan.from_saved)}, stay where they ` +
      "are: delete them yourself if you don't want them."
    : "";
  return ask("Use the records already there?",
    `isoshelf already has records for this folder in ${plan.to_dir}, saved ` +
    `${timeAgo(plan.to_saved)}. It will use those.` + left,
    [{ label: "Use those", value: true, primary: true }, cancel]);
}

// recordsDone says what the last move did, in a line.
function recordsDone(m) {
  switch (m.what) {
    case "move": return `Moved here from ${m.from}.`;
    case "keep": return `Using the records that were already here. The ones in ${m.from_dir} ` +
      "are still there - delete them yourself if you don't want them.";
    case "use": return "Using the records that were already here.";
    case "none": return "Nothing to move yet. This folder's records are kept here from now on.";
    default: return "";
  }
}

// recordsNote is the note under the setting: what the last move did, if it
// was for this folder, and then where the records are now.
function recordsNote() {
  const r = (state && state.records) || {};
  if (!r.file) return "";
  let where = `Right now: ${r.file}`;
  if (r.location !== "folder") {
    where += " - kept away from the folder, they're found by its path, so the same " +
      "drive at a different letter or mount point starts with nothing. Your images " +
      "and the archive never move.";
  }
  const done = recordsResult && recordsResult.target === state.target ? recordsDone(recordsResult) : "";
  return [done ? el("div", { class: "records-done", role: "status" }, done) : null, where];
}

// recordsControl is the three answers, plus the box for the third one. The
// first two ask as soon as they are picked; the third waits until a folder
// has been typed, because half a path is not a folder.
function recordsControl() {
  const now = (state && state.records) || {};
  const chosen = recordsChoice === null ? (now.location || "folder") : recordsChoice;
  const rows = [
    el("select", {
      "aria-label": "Where this folder's records are kept",
      onchange: (e) => {
        const value = e.target.value;
        if (value === "custom") {
          recordsChoice = "custom";
          renderSettings(true);
          return;
        }
        chooseRecords({ location: value });
      },
    }, [
      ["folder", "In the folder itself"],
      ["app", "In isoshelf's own folder"],
      ["custom", "In a folder you pick"],
    ].map(([value, text]) => el("option", { value, selected: value === chosen || undefined }, text))),
  ];
  if (chosen === "custom") {
    const box = el("input", {
      type: "text", class: "records-dir", value: now.dir || "",
      placeholder: "The whole path to a folder",
      "aria-label": "The folder to keep this folder's records in",
      onkeydown: (e) => {
        if (e.key === "Enter") chooseRecords({ location: "custom", dir: e.target.value });
      },
    });
    rows.push(box, el("button", {
      type: "button", class: "btn small",
      onclick: () => chooseRecords({ location: "custom", dir: box.value }),
    }, "Use this folder"));
  }
  return el("div", { class: "setting-controls" }, rows);
}
